lib/evaluate/host.h

```diff
@@ -0,0 +1,229 @@
+namespace host {
+
+// Type mapping from F18 types to host types
+struct UnsupportedType {};  // There is no host type for the F18 type
+
+template<typename FTN_T> struct HostTypeHelper {
+  using Type = UnsupportedType;
+};
+template<typename FTN_T> using HostType = typename HostTypeHelper<FTN_T>::Type;
+
+template<typename... T> constexpr inline bool HostTypeExists() {
+  return (... && (!std::is_same_v<HostType<T>, UnsupportedType>));
+}
+
+// Type mapping from host types to F18 types FortranType<HOST_T> is defined
+// after all HosTypeHelper definition because it reverses them to avoid
+// duplication.
+
+// Scalar conversion utilities from host scalars to F18 scalars
+template<typename FTN_T>
+inline constexpr Scalar<FTN_T> CastHostToFortran(const HostType<FTN_T> &x) {
+  static_assert(HostTypeExists<FTN_T>());
+  if constexpr (FTN_T::category == TypeCategory::Complex &&
+      sizeof(Scalar<FTN_T>) != sizeof(HostType<FTN_T>)) {
+    // X87 is usually padded to 12 or 16bytes. Need to cast piecewise for
+    // complex
+    return Scalar<FTN_T>{CastHostToFortran<typename FTN_T::Part>(std::real(x)),
+        CastHostToFortran<typename FTN_T::Part>(std::imag(x))};
+  } else {
+    return *reinterpret_cast<const Scalar<FTN_T> *>(&x);
+  }
+}
+
+// Scalar conversion utilities from  F18 scalars to host scalars
+template<typename FTN_T>
+inline constexpr HostType<FTN_T> CastFortranToHost(const Scalar<FTN_T> &x) {
+  static_assert(HostTypeExists<FTN_T>());
+  if constexpr (FTN_T::category == TypeCategory::Complex &&
+      sizeof(Scalar<FTN_T>) != sizeof(HostType<FTN_T>)) {
+    // X87 is usually padded to 12 or 16bytes. Need to cast piecewise for
+    // complex
+    return HostType<FTN_T>{CastFortranToHost<typename FTN_T::Part>(x.REAL()),
+        CastFortranToHost<typename FTN_T::Part>(x.AIMAG())};
+  } else {
+    return *reinterpret_cast<const HostType<FTN_T> *>(&x);
+  }
+}
+
+template<typename T> struct BiggerOrSameHostTypeHelper {
+  using Type =
+      std::conditional_t<HostTypeExists<T>(), HostType<T>, UnsupportedType>;
+  using FortranType = T;
+};
+
+template<typename FTN_T>
+using BiggerOrSameHostType = typename BiggerOrSameHostTypeHelper<FTN_T>::Type;
+template<typename FTN_T>
+using BiggerOrSameFortranTypeSupportedOnHost =
+    typename BiggerOrSameHostTypeHelper<FTN_T>::FortranType;
+
+template<typename... T> constexpr inline bool BiggerOrSameHostTypeExists() {
+  return (... && (!std::is_same_v<BiggerOrSameHostType<T>, UnsupportedType>));
+}
+
+// Defining the actual mapping
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 1>> {
+  using Type = std::int8_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 2>> {
+  using Type = std::int16_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 4>> {
+  using Type = std::int32_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 8>> {
+  using Type = std::int64_t;
+};
+
+// no int 128bit
+
+// no float 16bits
```

Comment:
Why not?

Reply:
It is implicitly mapped to UnsupportedType. There is no "half" in the standard yet. Are they any architecture where float is 16bits ? Then it may be worth adding a conditional that tries mapping it to float.

Reply:
The IEEE-754 2008 standard has `binary16`, and some environments have the alternative `bfloat16` which is just 32-bit floating-point without the least-order 16 bits.  f18 currently supports both, and the runtime library might have to support one or both.  I don't think that either form is supported by gcc or clang or by standard C math libraries.

Reply:
If gcc and clang are not supporting it right now, I propose transforming the comment regarding the host/f18 types mapping into a TODO for when gcc and clang supports it.
For folding purposes, we can convert to "bigger" supported floating-point. I have to find a good place to make the conversions.